packages/grid/x-data-grid/src/hooks/features/filter/gridFilterUtils.ts

```diff
@@ -234,20 +236,60 @@ export const buildAggregatedFilterItemsApplier = (
     return null;
   }
 
-  return (row, shouldApplyFilter) => {
-    const resultPerItemId: GridFilterItemResult = {};
-
-    for (let i = 0; i < appliers.length; i += 1) {
-      const applier = appliers[i];
-      if (!shouldApplyFilter || shouldApplyFilter(applier.item.field)) {
-        resultPerItemId[applier.item.id!] = applier.v7
-          ? applier.fn(row)
-          : applier.fn(getRowId ? getRowId(row) : row.id);
-      }
-    }
-
-    return resultPerItemId;
-  };
+  // Original logic:
+  // return (row, shouldApplyFilter) => {
+  //   const resultPerItemId: GridFilterItemResult = {};
+  //
+  //   for (let i = 0; i < appliers.length; i += 1) {
+  //     const applier = appliers[i];
+  //     if (!shouldApplyFilter || shouldApplyFilter(applier.item.field)) {
+  //       resultPerItemId[applier.item.id!] = applier.v7
+  //         ? applier.fn(row)
+  //         : applier.fn(getRowId ? getRowId(row) : row.id);
+  //     }
+  //   }
+  //
+  //   return resultPerItemId;
+  // };
+
+  // We generate a new function with `eval()` to avoid expensive patterns for JS engines
+  // such as a dynamic object assignment, e.g. `{ [dynamicKey]: value }`.
+  const filterItemTemplate = `(function filterItem$$(row, shouldApplyFilter) {
+      ${appliers
+        .map(
+          (applier, i) =>
+            `const shouldApply${i} = !shouldApplyFilter || shouldApplyFilter(${JSON.stringify(
+              applier.item.field,
+            )});`,
+        )
+        .join('\n')}
+
+      const result$$ = {
+      ${appliers
+        .map(
+          (applier, i) =>
+            `${applier.item.id !== undefined ? parseInt(String(applier.item.id), 10) : 'undefined'}:
```

Comment:
`applier.item.id` [could also be a string](https://github.com/mui/mui-x/blob/aa9186b0ba6ea1116b1f8048a3fb7f4072591e20/packages/grid/x-data-grid/src/models/gridFilterItem.ts#L11), so we should rather use `JSON.stringify` here.
Here's an example showing the issue: https://codesandbox.io/s/fervent-diffie-kl8zkm?file=/demo.tsx

Reply:
Done. Our item.id is also sometimes `undefined` :|